File: packages-logicn/logicn-devtools-benchmarks/benchmarks/collection-pipeline/bench.cpp
#include <cstdint>
#include <chrono>
#include <iostream>
#include <cstring>
#include <cstdlib>
#include <vector>
#include <numeric>

int64_t pipeline(const std::vector<int>& arr) {
    int64_t sum = 0;
    for (int x : arr) {
        if (x % 2 == 0) sum += (int64_t)x * 2;
    }
    return sum;
}

int main(int argc, char* argv[]) {
    int size = 10000;
    int its = 100000;
    for (int i = 1; i < argc - 1; i++) {
        if (!strcmp(argv[i], "--size")) size = atoi(argv[i+1]);
        if (!strcmp(argv[i], "--operations") || !strcmp(argv[i], "--iterations")) its = atoi(argv[i+1]);
    }
    std::vector<int> arr(size);
    std::iota(arr.begin(), arr.end(), 0);
    // Warmup
    pipeline(arr);
    auto t0 = std::chrono::steady_clock::now();
    int64_t result = 0;
    for (int i = 0; i < its; i++) { result = pipeline(arr); }
    double ms = std::chrono::duration<double, std::milli>(std::chrono::steady_clock::now() - t0).count();
    std::cout << "{\"runtime\":\"cpp\",\"benchmark\":\"collection-pipeline-v1\","
              << "\"size\":" << size << ","
              << "\"iterations\":" << its << ","
              << "\"result\":" << result << ","
              << "\"elapsedMs\":" << ms << ","
              << "\"iterationsPerSecond\":" << its / (ms / 1000) << "}" << std::endl;
}

File: packages-logicn/logicn-devtools-benchmarks/benchmarks/collection-pipeline/bench.rs
use std::time::Instant;
use std::env;

fn pipeline(size: usize) -> i64 {
    (0..size)
        .filter(|x| x % 2 == 0)
        .map(|x| (x * 2) as i64)
        .sum()
}

fn main() {
    let mut size: usize = 10_000;
    let mut iterations: usize = 100_000;
    let args: Vec<String> = env::args().collect();
    let mut i = 1;
    while i + 1 < args.len() {
        match args[i].as_str() {
            "--size"       => size       = args[i+1].parse().unwrap_or(size),
            "--operations" | "--iterations" => iterations = args[i+1].parse().unwrap_or(iterations),
            _ => {}
        }
        i += 2;
    }
    // Warmup
    let _ = pipeline(size);
    let t0 = Instant::now();
    let mut result: i64 = 0;
    for _ in 0..iterations {
        result = pipeline(size);
    }
    let elapsed = t0.elapsed().as_secs_f64() * 1000.0;
    println!(r#"{{"runtime":"rust","benchmark":"collection-pipeline-v1","size":{},"iterations":{},"result":{},"elapsedMs":{:.3},"iterationsPerSecond":{:.2}}}"#,
        size, iterations, result, elapsed, iterations as f64 / (elapsed / 1000.0));
}
